### ISSUE TITLE
JSON Waveforms

### ISSUE THREAD

**Issue by GianlucaCal** [laurentvd/wavesurfer.swf#6]
Hello,

is there a way to pass to the wavesurfer.load the waveforms with a JSON file? This fallback doesnt visualize long audio files so I was trying to fix it with the processing of waveform server-side.

Thanks

**Comment by laurentvd** [comment]
Not at the moment, but I guess it would be possible to build this. Is this something the HTML variant does support?

**Comment by GianlucaCal** [comment]
yes the HTML version does support it, [link](https://wavesurfer-js.org/example/audio-element/),

basically the syntax for this is `waveform.load('example.wav',JSON.parse(example.json))`

**Comment by laurentvd** [comment]
Alright, I just added support for passing your own waveform! Let me know if it works for you.

**Comment by GianlucaCal** [comment]
thank you so much for that, it seems to work! Excuse me if I go off topic but are there any chances to implement also the zoom and region selection functions?

**Comment by laurentvd** [comment]
You're welcome! About the zoom and region functionalities; I feel the functionalities should never have been added to the .js version. I guess they could be implemented, but for now I think it's too much work. That said, if someone would create a PR for it, I'd be happy to merge it.

**Comment by GianlucaCal** [comment]
Hi, I found an issue about the load with a prerendered waveform array, if the array is not normalize between -1 and 1 the waveform doesnt visualize correctly, is there an easy fix to it?

**Comment by laurentvd** [comment]
It is certainly possible, but I'm not sure it is the expected behaviour. I mean, if I look at the .js example at [https://wavesurfer-js.org/example/audio-element/main.js](https://wavesurfer-js.org/example/audio-element/main.js) there are no negative values and all values are between `0` and `1`. What would a negative value mean? And what do you expect to happen if a value exceeds `1`?

**Comment by GianlucaCal** [comment]
When the values of the array exceed `1` the drawer saturate the Canvas. So since the waveform command line, used to retreive the JSON in the HTML version of wavesurfer, doesnt return values  normalized between `0` and `1` [(waveform image generator)](https://github.com/bbc/audiowaveform) the developer implemented an option which normalize the values. [Wavesurfer options ](https://wavesurfer-js.org/docs/options.html), can you help me to implement the _normalize option_ on the fallback? Moreover seems impossible to load the array from a JSON retrieved from a remote URL!

**Comment by laurentvd** [comment]
I see. I haven't touched Waveform.js in a while, so wasn't aware of the new cool features it adds. Apparently it *scales* the values by default, but *normalize* scales to the highest peak value. We do neither at the moment.

The scaling shouldn't be too hard, but normalization might be. It could be hard to keep it fast. Would you like to give it a go? If so, I suggest you make PR for [AudioData.as](https://github.com/laurentvd/wavesurfer.swf/blob/master/src/fm/wavesurfer/audio/AudioData.as).

**Comment by GianlucaCal** [comment]
I'll try to give it a look! What about the loading of the array of peaks from a JSON retrieved from a URL? Do you think it's possible to implement it?

**Comment by laurentvd** [comment]
Definitely possible, I'll have a look at it later this week. Also ActionScript is not my day-to-day language so it always take a bit of time ;-)
